### PR TITLE
Promo renew

### PR DESCRIPTION
**Pull request by johnduffell**
This allows promos on renew.  So far it works with discounts, but I'm struggling to test with custom created promotions.  I will carry on trying to test with free trials and tracking

bugs to be fixed are:
- promo code isn't getting written to the subscription - need to find the right kind of Amend call in Commands.scala as it's not TermsAndConditions
- the client side code assumes your zuora account is using the default currency of your billing country, which means promotions are displayed in dollars if you are paying in pounds with a US billing address

@pvighi @paulbrown1982 @AWare

### CODE REVIEW

File: app/controllers/Promotion.scala [hunk 1]
```diff
@@ -65,7 +65,7 @@ object Promotion extends Controller with LazyLogging with CatalogProvider {
         "isValid" -> result.isRight,
         "errorMessage" -> result.swap.toOption.map(_.msg)
       )
-      result.fold(_ => NotAcceptable(body), _ => Ok(body))
+      result.fold(_ => NotAcceptable/*FIXME prob 404*/(body), _ => Ok(body))
```
Comment (paulbrown1982) [inline app/controllers/Promotion.scala]: If this needs to be fixed before go-live, please can you raise a Trello ticket?

File: app/services/CheckoutService.scala [hunk 2]
```diff
@@ -280,10 +281,17 @@ class CheckoutService(identityService: IdentityService,
...
-    def addPlan = {
+    def addPlan(contact: Contact) = {
       val newRatePlan = RatePlan(renewal.plan.id.get, None)
       val renewCommand = Renew(subscription.id.get, subscription.startDate, NonEmptyList(newRatePlan), contractEffective, customerAcceptance)
-      zuoraService.renewSubscription(renewCommand)
+      val validPromotion = for {
+        code <- renewal.promoCode
+        deliveryCountryString <- contact.mailingCountry
+        deliveryCountry <- CountryGroup.countryByName(deliveryCountryString)
+        validPromotion <- promoService.validate[NewUsers](code, deliveryCountry, renewal.plan.id).toOption // FIXME this line to have 3 states
```
Comment (paulbrown1982) [inline app/services/CheckoutService.scala]: If this needs to be fixed before go-live, please can you raise a Trello ticket?

File: app/services/CheckoutService.scala [hunk 3]
```diff
@@ -80,7 +80,7 @@ class CheckoutService(identityService: IdentityService,
       defaultPaymentDelay = CheckoutService.paymentDelay(subscriptionData.productData, zuoraProperties)
       paymentMethod <- EitherT(attachPaymentMethodToStripeCustomer(payment, purchaserIds))
       subscribe = createSubscribeRequest(personalData, soldToContact, requestData, plan, purchaserIds, paymentMethod, payment.makeAccount, Some(defaultPaymentDelay))
-      validPromotion = promoCode.flatMap(promoService.validate[NewUsers](_, personalData.address.country.getOrElse(Country.UK), subscriptionData.productRatePlanId).toOption)
+      validPromotion = promoCode.flatMap(promoService.validate[PromoContext.Renewal](_, personalData.address.country.getOrElse(Country.UK), subscriptionData.productRatePlanId).toOption)
```
Comment (paulbrown1982) [inline app/services/CheckoutService.scala]: For Guardian Weekly subs (acquisition AND renewal), the country being validated for the promotion should be the delivery country not the billing country.

Happy if this is created as a new ticket.

File: app/controllers/Promotion.scala (hunk 1 above)
Comment (johnduffell) [inline app/controllers/Promotion.scala]: no, it's just an inappropriate http code response for an invalid code

File: app/services/CheckoutService.scala [hunk 4]
```diff
@@ -280,10 +281,17 @@ class CheckoutService(identityService: IdentityService,
...
-    def addPlan = {
+    def addPlan(contact: Contact) = {
       val newRatePlan = RatePlan(renewal.plan.id.get, None)
       val renewCommand = Renew(subscription.id.get, subscription.startDate, NonEmptyList(newRatePlan), contractEffective, customerAcceptance)
-      zuoraService.renewSubscription(renewCommand)
+      val validPromotion = for {
+        code <- renewal.promoCode
+        deliveryCountryString <- contact.mailingCountry
+        deliveryCountry <- CountryGroup.countryByName(deliveryCountryString)
+        validPromotion <- promoService.validate[NewUsers](code, deliveryCountry, renewal.plan.id).toOption // would be better to have valid/invalid/tracking
```
Comment (paulbrown1982) [inline app/services/CheckoutService.scala]: Agreed, we should do the conversion to convert the promotion into a Tracking promo code if it's 'valid' but not for the chosen rate plan.

Comment (johnduffell) [inline app/services/CheckoutService.scala]: ok I will do that next

**Pull request merged**